golem/managers/payment/pay_all.py

```diff
@@ -30,116 +34,119 @@ def __init__(
         budget: float,
         network: str = DEFAULT_PAYMENT_NETWORK,
         driver: str = DEFAULT_PAYMENT_DRIVER,
+        shutdown_timeout: timedelta = DEFAULT_SHUTDOWN_TIMEOUT,
     ):
         self._golem = golem
         self._budget = budget
         self._network = network
         self._driver = driver
+        self._shutdown_timeout = shutdown_timeout.total_seconds()
 
         self._allocation: Optional[Allocation] = None
 
-        self._opened_agreements_count = 0
-        self._closed_agreements_count = 0
-        self._payed_invoices_count = 0
-
         self._event_handlers: List = []
 
+        self._agreements: Dict[str, Agreement] = {}
+        self._agreement_event: asyncio.Event = asyncio.Event()
+
     @trace_span("Starting PayAllPaymentManager", log_level=logging.INFO)
     async def start(self):
-        # TODO: Add stop with event_bus.off()
         self._event_handlers.extend(
             [
                 await self._golem.event_bus.on(NewInvoice, self._pay_invoice_if_received),
                 await self._golem.event_bus.on(NewDebitNote, self._pay_debit_note_if_received),
-                await self._golem.event_bus.on(NewAgreement, self._increment_opened_agreements),
-                await self._golem.event_bus.on(AgreementClosed, self._increment_closed_agreements),
+                await self._golem.event_bus.on(NewAgreement, self._handle_new_agreement),
             ]
         )
 
-    @trace_span("Stopping PayAllPaymentManager", log_level=logging.INFO)
-    async def stop(self):
-        try:
-            await self.wait_for_invoices()
-        except ManagerException:
-            pass
+    @trace_span("Getting allocation", show_results=True, log_level=logging.INFO)
+    async def get_allocation(self) -> Allocation:
+        if self._allocation is None:
+            self._allocation = await self._create_allocation()
 
-        for event_handler in self._event_handlers:
-            await self._golem.event_bus.off(event_handler)
+        return self._allocation
+
+    async def _release_allocation(self) -> None:
+        if self._allocation is None:
+            return
+
+        await self._allocation.release()
```

Comment:
`self._allocation` should be defaulted back to the `None` value after release to avoid a stale state for potential manager restarts.